### PR TITLE
Add VM Definition proposal to the vm-configuration proposal

### PR DESCRIPTION
**Pull request by mpolednik**
Signed-off-by: Martin Polednik <mpolednik@redhat.com>

### CODE REVIEW

File: docs/vm-configuration.md [hunk 1]
```diff
@@ -264,3 +264,95 @@ the generation of multiple libvirt XML documents, the domain XML and associated
...
+create modular VM Spec generation flows that are easily tunable by experts in
+their domains of expertise.
+
+```yaml
+apiVersion: kubevirt.io/v1alpha2
+kind: VM
+metadata:
+  name: testvm
+spec:
+  features:
```
Comment (fabiand) [inline docs/vm-configuration.md]: How do you handle conflicts between `features` and an eventual `domain` specification?

Comment (fabiand) [inline docs/vm-configuration.md]: Could you explain why you would add the `features` section to the `VM` object rather than to the `VMConfig` object?

Comment (mpolednik) [inline docs/vm-configuration.md]: This is something I'd love to have open discussion about. My vision for features section is that it will be *the* user specified configuration, while domain (VM Spec) section is more of a runtime, low level description. If we generate VM Spec from features, there should be no conflict. If user directly modifies VM Spec, it should be considered power user task that may lead to inconsistencies (think of it as hooking mechanism in VDSM).

Comment (mpolednik) [inline docs/vm-configuration.md]: Not everyone will be using VMConfig, forcing management applications that store VMs on their own to use VMConfig is redundant.

Comment (michalskrivanek) [inline docs/vm-configuration.md]: Not sure if it is like that, but the earlier proposal for VMConfig vs VM would be close to the separation of high level representation and expanded domxml-like part. Wouldn't it then rather be that "noone" is using VM and users just always use VMConfig?

Comment (fabiand) [inline docs/vm-configuration.md]: Yes, `VMConfig` was intended to be a higher-level representation and a way to store non-running VMs.

A thought is to think about:

```yaml
kind: VMConfig
spec:
  features:
    .. # as described in this proposal
  domain:
    .. # to override stuff as needed

# THIS IS THEN TRANSFORMED on /start to a VM OBJ

kind: VM
spec:
  domain:
    ..
```

Thus `features` would just be a property of the `VMConfig` object, and be translated during start into the domain.

Comment (mpolednik) [inline docs/vm-configuration.md]: If no user (where user is anything that wishes to use kubevirt, incl. applications) is expected to deal with plain VM objects, this would work. As far as I remember what @rmohr said, VMConfig is just another way of interfacing kubevirt - in that case, having features *only* in VMConfig would lead to different user experience wrt. VM definition depending on which mechanism the user decides to use. That is confusing and may force users to use specific, opinionated way of defining VMs.

Comment (michalskrivanek) [inline docs/vm-configuration.md]: @mpolednik well, surely the experience is different, VMConfig is an abstraction of details in VM. It's not that VMConfig features cannot be expressed in VM, rather it's just too low level/convoluted way to express them in VM.

Comment (mpolednik) [inline docs/vm-configuration.md]: It's not only abstraction but also a persistent definition -  similar to persistent libvirt domains. Using VMConfig with custom persistent storage will lead to 3 forms of persistence used:

1. App specific persistence,
2. VMConfig persistence,
3. persistent libvirt domain.

We should decide on at most two (one being libvirt) persistent layers.

Comment (mpolednik) [inline docs/vm-configuration.md]: After offline discussion, I believe that having features as a part of VMConfig and advertising VMConfig as *the* way to interact with KubeVirt is the most reasonable solution.

Comment (rmohr) [inline docs/vm-configuration.md]: While I don't think that something like a feature section should be on the VM, I would not promote VMConfig as the main entrypoint to KubeVirt.

Comment (rmohr) [inline docs/vm-configuration.md]: Could we share the reasoning?

Comment (rmohr) [inline docs/vm-configuration.md]: What exactly is the advantage, over just specifying in the spec, what you care about, and let the reset be filled in?

Did you also think about some higher features? There was for instance some discussion, about flavours, ... which could be implemented via admission controllers.

Could you also share, how the same sections look like in the libvirt xml, so that we better see the benefit of this feature section?

Comment (rmohr) [inline docs/vm-configuration.md]: Regarding the conflict resolution, [podpresets](https://kubernetes.io/docs/tasks/inject-data-application/podpreset/) have to deal with that too. Maybe that is of interest. If such features (in a little bit more high level form), would be implemented via a mapping done by admission controllers, havein skip policies on such admission controllers could solve conflict scenarios (e.g. Optional, RejectIfConflicting, ...)

Comment (rmohr) [inline docs/vm-configuration.md]: If these features would be a little bit more high level,  and done with some kind of admission control mechanism (so that we still have the one and only spec at the end), it would imho go into the right direction. Just this kind of separate high level spec in the same document does not sound like the right abstraction for me.

Comment (fabiand) [inline docs/vm-configuration.md]: Maybe we can carefully phrase it as, _use VM for full controll, or VMConfig for ease and persistence_. Something along those lines.

@rmohr The "classic" admission controllers modify the spec during creation (and thus store a modified spec), to me this is suboptimal for this `feature` feature (ha!).

If a `feature` is specified, then we still need to be able to see in the retrospective, which feature it was. In this two-spec case, it's easy and straight forward.
Technically we could probably push this into the VM spec, but I've got the impression that it will get messy.

To me the picture has three levels:
1. private/internal, lowest level, host specific libvirt domain xml, can not be accessed from the outside
2. public cluster wide vm domain representation (`VM`) could also be called an instance
3. public cluster wide abstract and persistent vm representation (`VMConfig`)

an instance (`VM`) would always (if it's used) be derived from a `VMConfig´, and this - IMO - makes it easy to understand the transitions, as well as understanding, from which high-level definition a VM definition got derived.

Comment (rmohr) [inline docs/vm-configuration.md]: Yes, no objections. Just keep in mind, that admission controllers can modify the spec based on something like such features always, since the original spec was posted from somewhere. The poster always has the old not-yet-expanded/defaulted spec to compare it with the runtime one. I just want to make sure, that we don't exclude the runtime from usefull featurs, like libosinfo admission controll too lighthearted.

Comment (michalskrivanek) [inline docs/vm-configuration.md]: @fabiand,@rmohr: How about looking at it as:

kind: VMConfig
spec:
  features:
    .. # as described in this proposal

# THIS IS THEN TRANSFORMED on /start to a VM OBJ
kind: VM
  features:
    .. # as described in this proposal
  domain:
    .. # resulting detailed structure, consumed by virt-handler

- domain resembles libvirt XML in structure (all detailed devices, PCI addresses?, timers, etc), it is further transformed by virt-handler to resulting libvirt XML, but that expansion is only simple straightforward one (like adding kubevirt-hardcoded options, e.g. "os type hvm", etc) and plus all the host-specific things (actual mapping of the backend resources)
- the domain part is "generated" by the admission controller of the VM entity, that's where all the logic of how the VM looks like resides, based on e.g. OS and features like "use virtio-scsi".
- there is a duplicity between features and domain section in VM. You can basically either post only "features"(high level) or only "domain"(low level description) or both, but in that case you need to know about potential conflicts in implementation of each "feature->domain" in the admission controller. Still, it seems to me as the lesser of all other evils.

doesn't it solve the problem of the main entry point? This way interacting with VM still gives you full power and the logic is kept in single place, VMConfig is only for persistence, separate from runtime.

Comment (rmohr) [inline docs/vm-configuration.md]: If we really say, that the `feature` section allows a full description, then it is probably not so attractive to have the feature section on the `VM` object too. Then I would just use `VMConfig` and store the VMSpec for VM restarts on the VMConfig. Even though reusing as much as possible in the context of admission controllers would be something I love I would then try to avoid the conflict situations right now. If we would also allow on the `VMConfig` to set values in `feature` and `domain`, then I would  tend to having something like these `features` on the `VM`, since we already have to deal with  conflicts anyway, and the runtime could profit from it. E.g. if we introduce something like a `VirtualMachineReplicationController`, it could still work with the `VM`, and would not have to go through `VMConfig`, or have a `VirtualMachineReplicationControllerConfig` to provide it's features.

File: docs/vm-configuration.md [hunk 2]
```diff
@@ -264,3 +264,312 @@ the generation of multiple libvirt XML documents, the domain XML and associated
...
+    graphics: true # graphics, video and spice channel (or VNC settings for different architecture) sections with reasonable defaults
+    system:
+      type: windows # adds guest agent channel
+      version: 10 # generates timers and features sections
+      arch: x86_64 # enough for os section
+    storage:
+      - type: cdrom
+      - type: disk # duplicates some of the data in spec but eliminates the need for low level details
+        interface: virtio-scsi # also generates virtio-scsi controller
+        path: /mntpoint/11111111-1111-1111-1111-111111111111/images/22222222-2222-2222-2222-222222222222/22222222-2222-2222-2222-222222222222
```
Comment (mpolednik) [inline docs/vm-configuration.md]: That is wrong, host-specific stuff shouldn't be included here.

File: docs/vm-configuration.md [hunk 3]
```diff
@@ -264,3 +264,312 @@ the generation of multiple libvirt XML documents, the domain XML and associated
...
+          startupPolicy: optional
+        target:
+          bus: ide
+          dev: hdc
+        readonly: true
+      - device: disk 
+        snapshot: false
+        type: file
+        source:
+          file: /mntpoint/11111111-1111-1111-1111-111111111111/images/22222222-2222-2222-2222-222222222222/22222222-2222-2222-2222-222222222222
```
Comment (ahadas) [inline docs/vm-configuration.md]: I think that even in the "lower level" representation, the VM configuration should not point to the the exact file of the disk. Let's say that we do live storage migration then we initially create a snapshot on top of each disk - such representation would make us change the VM configuration at this point and later on when merging the temporary snapshot. I think that a better way is pointing to the disk (like in line 546), assuming that the leaf volume is used.

File: docs/vm-configuration.md [hunk 4]
```diff
@@ -264,3 +264,312 @@ the generation of multiple libvirt XML documents, the domain XML and associated
...
+    network:
+      - source: mgmtnet
+        mac: 00:aa:bb:cc:dd:ee
+    rng: true
+  domain:
+    <spec section of the yaml snippet above>
+```
+
+The features section is part of a VM Definition, therefore it is also suitable
+for inclusion in VMConfig. Accessing domain part of the spec is considered
```
Comment (ahadas) [inline docs/vm-configuration.md]: One of the arguments for VMConfig I read in a previous discussion was "for the persistence of non-running VMs" - but it doesn't hold pci addresses, right? so we will probably need to persist the lower-level configuration as well. I wonder if defining a reasonable mapping would be better than persisting two different representations of the same entity.

File: docs/vm-configuration.md [hunk 5]
```diff
@@ -264,3 +264,312 @@ the generation of multiple libvirt XML documents, the domain XML and associated
...
+         source:
+           mode: bind
+           path: /var/lib/libvirt/qemu/channels/11111111-feed-dead-beef-111111111111.org.qemu.guest_agent.0
+         target:
+           name: org.qemu.guest_agent.0
+           type: virtio
+       - type: spicevmc
+         target:
+           name: com.redhat.spice.0
+           type: virtio
```
Comment (ahadas) [inline docs/vm-configuration.md]: is it intentional that pci addresses are missing?

**Comment by fabiand** [comment]
Seems that host specific bits slipped into the `VM` object example. But The `VM` obj must not contain any host specific paths.

@ahadas In another thread there was a discussion if pci addresses are really needed, or if another mechanism to identify devices directly is also suited.

**Comment by ahadas** [comment]
@fabiand , is there a need for the paths to be host-specific?

File: docs/vm-configuration.md (hunk 4 above)
Comment (mpolednik) [inline docs/vm-configuration.md]: That may not be needed in the end -- e.g. if device ordering is stable we should be able to expect stable addresses..

File: docs/vm-configuration.md (hunk 3 above)
Comment (mpolednik) [inline docs/vm-configuration.md]: That is unrelated to the proposal, please see the storage proposals to see how if it could look in the end.

**Comment by mpolednik** [comment]
@fabiand Eveyrthing in the PR is an example, and storage isn't properly designed and implemented yet. These points can be fixed any time we have a final storage design.

**Comment by fabiand** [comment]
> @fabiand , is there a need for the paths to be host-specific?

We don't want absolute paths at all.

The VM API can contain references to host resources, but those host resources must be abstracted (see https://docs.google.com/document/d/1-BAR-RSFkXLbqRREW6jE7JGipmGi2ygWxbwIZ8Eq_NQ/edit#).

The abstractions ease two things (at least):
- Make the scheduler aware of resource requirements (i.e. a path could be a request to get access to an nfs share, or a hardwar rng)
- Keep the VM spec host unspecific, thus it can be handled by any host

The abstractions will then be resolved on the host by the handler, who will then i.e. replace them with paths to devices (i.e. /dev/random or some mount point).

**Comment by fabiand** [comment]
> @fabiand Eveyrthing in the PR is an example, and storage isn't properly designed and implemented yet. These points can be fixed any time we have a final storage design.

Alright, but the examples should still align with what we already know. One of it is that we must not have absolute file paths :)

**Comment by mpolednik** [comment]
Noted, will fix in next iteration. Let's try to keep the discussion on topic though.

**Comment by ahadas** [comment]
@fabiand ok

File: docs/vm-configuration.md [hunk 6]
```diff
@@ -264,3 +264,312 @@ the generation of multiple libvirt XML documents, the domain XML and associated
...
+This is an example of features section describing the VM Specification above:
+
+```yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VM
+metadata:
+  name: VM
+spec:
+  features:
+    memory: 4096 # other memory attributes can be deducted
```
Comment (MarSik) [inline docs/vm-configuration.md]: This is an oversimplification. You will definitely need current, minimal and maximal memory size. And we should probably use the same quantity format k8s uses (string with units eg. "10G")

Comment (michalskrivanek) [reply docs/vm-configuration.md]: sure, it's just an example. It should also include hugepages backing configuration, min mem guaranteed

File: docs/vm-configuration.md (hunk 4 above)
Comment (MarSik) [reply docs/vm-configuration.md]: I also do not see any QoS definitions here - libvirt exposes quite a few like cpu shares, IO limits, network limits...

Comment (fabiand) [reply docs/vm-configuration.md]: We aim at making the VM API to lead to stable hardware ABIs.

Thus - The same VM object should lead to the same domxml over time.

**Comment by rmohr** [comment]
Any updates on this? Is there still a feeling that we should discuss if such a simplified spec should be added to  the VM object? If not, the  document this is added seems to be the wrong spot then. Can we close this PR or will it be updated and the proposal moved somewhere else?

File: docs/vm-configuration.md [hunk 7]
```diff
@@ -264,3 +264,312 @@ the generation of multiple libvirt XML documents, the domain XML and associated
...
+    cpu:
+      topology:
```
Comment (fromanirh) [inline docs/vm-configuration.md]: why do we need to specify the topology? I think this could be deducted (with documented rules) and specified only if needed

**Comment by fabiand** [comment]
For the record, there will be a f2f tomorrow (see https://calendar.google.com/calendar/embed?src=18pc0jur01k8f2cccvn5j04j1g%40group.calendar.google.com&ctz=Etc%2FGMT) about this topic.

File: docs/vm-configuration.md (hunk 7 above)
Comment (michalskrivanek) [reply docs/vm-configuration.md]: Many of those are not required to be specified, they come from templating/defaulting/expanding(pick your favorite term:) which is out of scope of this particular proposal.

**Comment by davidvossel** [comment]
I'm going to close this discussion for now. Feel free to re-open it if there's more we should talk about.

The resolution over the past few months is that the virtualmachine api will fairly closely match libvirt's domain xml.  However, The virtual machine api will be a super set of the libvirt xml and some kubevirt specific features (like registry disks and cloud init support).  Essentially, we will have the ability to express anything libvirt is capable of and more in the kubevirt api.

The persistent virtual machine config object is some that is being developed in the kubevirt-incubator project.
